rad/src/lib.rs

```diff
@@ -349,40 +357,118 @@ fn add_http_headers(
+/// HTTP transports at once.
+pub async fn run_paranoid_retrieval(
+    retrieve: &RADRetrieve,
+    tally: &RADTally,
+    settings: RadonScriptExecutionSettings,
+    active_wips: &ActiveWips,
+    transports: &[Option<String>],
+) -> Result<RadonTypes> {
+    let futures: Result<Vec<_>> = transports
+        .iter()
+        .map(|transport| {
+            WitnetHttpClient::new(transport)
+                .map_err(|err| RadError::HttpOther {
+                    message: err.to_string(),
+                })
+                .map(|client| {
+                    run_retrieval_report(
+                        retrieve,
+                        RadonScriptExecutionSettings::disable_all(),
+                        active_wips,
+                        Some(client),
+                    )
+                })
+        })
+        .collect();
+
+    let values = join_all(futures?).await;
+
+    evaluate_paranoid_retrieval(values, tally, settings)
+}
+
+/// Evaluate whether the values obtained when retrieving a data source through multiple transports
+/// are consistent, i.e. they all pass the filters from the tally stage.
+fn evaluate_paranoid_retrieval(
+    data: Vec<Result<RadonReport<RadonTypes>>>,
+    tally: &RADTally,
+    settings: RadonScriptExecutionSettings,
+) -> Result<RadonTypes> {
+    let mut values = vec![];
+
+    // Turn the reports into values, short-circuiting if any retrieval inevitably failed.
+    for each in data {
+        let report = each?;
+        let result = report.into_inner();
+        values.push(result);
+    }
+
+    // If there was only one retrieved value, there's no actual need to run the tally, as this means
+    // that only one transport was used and therefore the node is not in paranoid mode.
+    if values.len() > 1 {
+        let mut context = ReportContext::from_stage(Stage::Tally(TallyMetaData::default()));
+        let _tally = run_tally_with_context_report(values.clone(), tally, &mut context, settings);
+
+        // If at least one of the retrieved values is marked as an outlier, resolve to the
+        // `InconsistentSource` error.
+        if let Stage::Tally(TallyMetaData { liars, .. }) = context.stage {
+            if !liars.is_empty() {
+                return Ok(RadonTypes::RadonError(RadonError::<RadError>::new(
+                    RadError::InconsistentSource,
+                )));
+            }
+        }
+    }
+
+    // If all values pass the filters, return the first value (any of them will work, actually)
+    Ok(values
+        .get(0)
+        .expect("At least one value must have been received")
+        .clone())
+}
+
 /// Run aggregate stage of a data request, return a tuple of `Result<RadonReport>` and `ReportContext`
 pub fn run_aggregation_report(
     radon_types_vec: Vec<RadonTypes>,
     aggregate: &RADAggregate,
     settings: RadonScriptExecutionSettings,
-    active_wips: ActiveWips,
+    active_wips: &ActiveWips,
```

Comment:
Gotcha. It is always a best practice to let the caller decide when to clone. Removing the `&` here will essentially displace the cloning responsibility to `<RadManager as Handler<ResolveRA>>::handle` (as the `Aggregate` structure is still needed further down the line, beyond the lifetime of `run_paranoid_retrieval`, namely at `run_aggregation_report`).

Reply:
When I see stuff like this happen, I feel the urge to wrap every fricking property of a `struct` in an `Option`, then do `.take()` on the inner value to avoid any cloning :rofl:

Reply:
At a second thought... maybe the compiler is smart enough to realize that the value will be dropped shortly after the borrow and will actually skip the cloning for us? Or am I expecting just too much from it? :thinking:

Reply:
In any case, I have reverted these changes so `RADAggregate` are borrowed and the cloning happens in the outermost unit of business logic (`<RadManager as Handler<ResolveRA>>::handle`).

Reply:
> maybe the compiler is smart enough to realize that the value will be dropped without further usage shortly after the move of the clone and will actually skip the cloning for us?

Probably not because `ActiveWips` has a `HashMap` field, in theory we should be able to change it into an `Arc<HashMap>` and make the clone cheap, but not now.